### ISSUE TITLE
Add support for Java module path scanning (Jigsaw)

### ISSUE THREAD

**Issue by lukehutch** [lukehutch/fast-classpath-scanner#36]
As a benefit, this would be a precursor to supporting Ceylon.

**Comment by lukehutch** [comment]
This might be as simple as supporting the new Java 9 system classloaders, which are no longer instances of URLClassLoader. See:

http://openjdk.java.net/jeps/261

**Comment by lukehutch** [comment]
Will also need to search the codebase for references to `rt.jar` -- this jar doesn't exist in Java 9. Specifically:

* If `"!!"` is provided to the scan spec, something different will have to be done to include system jars in the scanning process.
* The code in JarUtils for testing whether a jar is a system jar will need to change.

**Comment by lukehutch** [comment]
See [https://docs.oracle.com/javase/9/migrate/toc.htm](https://docs.oracle.com/javase/9/migrate/toc.htm)

In particular, `rt.jar` no longer exists; neither does the `ext` dir; and in general, "jarfiles don't exist" in a compiled project (not sure what that means yet).

**Comment by lukehutch** [comment]
May need to provide support for [ModuleFinders](http://download.java.net/java/jdk9/docs/api/java/lang/module/ModuleFinder.html), not just ClassLoaders?

**Comment by lukehutch** [comment]
The universe of *observable modules* can be visualized with [java9-jigsaw-depviz](https://github.com/accso/java9-jigsaw-depvis). It looks like the code [here](https://github.com/accso/java9-jigsaw-depvis/blob/master/depvis/src/depvis/depvis/JigsawDepPrinter.java) is what is needed, e.g. `ModuleFinder.ofSystem().findAll()`. Will need to check if modRef has a URL field.

**Comment by lukehutch** [comment]
It may make sense to make Java 8 a minimum requirement when supporting Java 9 -- if so, the Java 9 changes should be made after the major API rework, so that the API rework is still available to Java 7 users in an older branch.

**Comment by lukehutch** [comment]
For JDK9, may need to read package name from constant pool rather than from file path (#141). (Module can also be stored there.)

**Comment by lukehutch** [comment]
[A Java 9 migration guide](https://blog.codefx.org/java/java-9-migration-guide/).

**Comment by lukehutch** [comment]
[Info on using Java 9 with maven-compiler-plugin](https://maven.apache.org/plugins/maven-compiler-plugin/examples/module-info.html) (N.B. maven-compiler-plugin needs to be at least version 3.7.0).

**Comment by lukehutch** [comment]
Added basic Java 9 classloader support in cf85f258. (See the TODO comment there.)

**Comment by johnou** [comment]
@lukehutch can you expand on "maven-compiler-plugin needs to be at least version 3.7.0"?

**Comment by lukehutch** [comment]
@johnou see ["Minimum compatible version" column here](https://cwiki.apache.org/confluence/display/MAVEN/Java+9+-+Jigsaw).

**Comment by johnou** [comment]
Perfect, thanks!

**Comment by lukehutch** [comment]
TODO: We're currently not parsing `module-info.class` at all, see these lines in `ClassfileBinaryParser.java`: 

```java
        if (isModule) {
            return null;
        }
```

I'm not sure yet what special parsing support will be needed for `module-info.class`.

**Comment by lukehutch** [comment]
[JDK 9 is strongly enforcing encapsulation](https://stackoverflow.com/a/41265267). One effect of this is that we can't get the `URLClassPath ucp` field from `AppClassLoader` or `BaseClassLoader`. This probably means we have to rely on JDK 9's own module system for scanning.

**Comment by lukehutch** [comment]
For the system modules:

```java
		// "Returns a module finder that locates the system modules. The system modules are the
		// modules in the Java run-time image. The module finder will always find java.base."
		Set<ModuleReference> ms = ModuleFinder.ofSystem().findAll();
		for (ModuleReference m : ms) {
			System.out.println("  " + m.descriptor().name());
			System.out.println("  " + m.descriptor().toNameAndVersion());
			System.out.println("  " + m.descriptor().packages());
			System.out.println("  " + m.descriptor().exports());
			Optional<URI> location = m.location();
			if (location.isPresent()) {
				System.out.println("  " + location.get()); // e.g. "jrt:/java.base"
			}
			m.open().list().forEach(s -> System.out.println("    " + s));
		}
```

I haven't figured out how to get the module list for non-system modules yet...

**Comment by lukehutch** [comment]
[Project Jigsaw: Module Class Loading and Bootstrapping -- Jigsaw Team](http://openjdk.java.net/projects/jigsaw/doc/module-class-loading.pdf)

**Comment by lukehutch** [comment]
See some discussion on related issues here:

https://github.com/policeman-tools/forbidden-apis/issues/54 (the code for `ReadableModulesResourceLoader` may be useful here)

https://github.com/policeman-tools/forbidden-apis/pull/95

**Comment by lukehutch** [comment]
Some good info on modules in Java 9:
https://blog.andresteingress.com/2017/09/29/java-9-modules.html

**Comment by lukehutch** [comment]
N.B. for reflection in Java 9+, will need to make FCS its own module, then open modules to the FCS module. From the Javadoc for `Module#addOpens(String packageName, Module other)`:

> If this module has opened a package to at least the caller module then update this module to open the package to the given module. Opening a package with this method allows all types in the package, and all their members, not just public types and their public members, to be reflected on by the given module when using APIs that support private access or a way to bypass or suppress default Java language access control checks.
> This method has no effect if the package is already open to the given module.

**Comment by lukehutch** [comment]
I sent the following email to the JPMS mailing list...

--

I am the author of Fast Classpath Scanner. I am trying to extend this library to work with JDK 9, and I'm running into a lot of apparent shortcomings of the JPMS API.

These are the main issues I have come across so far:

(1) Automatic modules from the legacy classpath cannot be found using `ModuleFinder.ofSystem()` (because they are not system modules), or using `ModuleLayer.boot().modules()` (because this list does not include unnamed modules).

(2) I would guess then that unnamed modules are in their own layer, but the Layer API only allows you to look at parent layers, not child layers -- so it is not even possible to enumerate all layers in the system. Why is there no call for enumerating all layers, and/or for reading child layers?

(3) `ModuleLayer.boot().configuration().findModule(name)` cannot find unnamed automatic modules either, because name cannot be null (and using the empty string doesn't return unnamed layers).

(4) From within a class in a traditional (non-module) jar on the classpath, using `getClass().getModule()`, does return a `Module` reference for the unnamed module. However, there is no API that I can find for getting a `ModuleReference` from a `Module` object, if the module is unnamed. For example, the following should work if the module is named, but does not work if the module is unnamed:

```java
    Optional<ModuleReference> moduleReference =
        module
            .getLayer()
            .configuration()
            .findModule(module.getName())
            .map(ResolvedModule::reference);
```

(5) `ModuleFinder.of(path)` would find the module, if the path were known. But there is no way to get the path from the Module reference obtained using `getClass().getModule()` -- you need a `ModuleReference` object to get the `URI` of the module, and you cannot get a `ModuleReference`, as described in (4) above.

(6) Having to specify a path when calling `ModuleFinder.of()`, as shown in (5), defeats the purpose of having a classpath and/or module path that is specified outside of program code. I imagine it is primarily useful for dynamic loading of modules, but what I'm interested in is the static classpath / module path.

(7) Module has a method `Set<String> getPackages()`, which allows you to enumerate package names within a module, but it does not have a method for listing resource names within each package. For that, again you need a `ModuleReference`, and then you can call `ModuleReference#open()` to get a `ModuleReader`, then `ModuleReader#list()` to list the resources in the module. Why is there no `open()` method in `Module`, and/or why is it not possible to get a `ModuleReference` directly from a `Module`?

It seems like there are lots of obvious omissions in the API, from an enumeration / scanning point of view. Or am I missing something?

**Comment by lukehutch** [comment]
The link to this message in the jigsaw-dev archives. Nobody has replied yet; hopefully somebody will provide some answers after the Christmas / New Year's break.

http://mail.openjdk.java.net/pipermail/jigsaw-dev/2017-December/013439.html

**Comment by lukehutch** [comment]
@HoldYourWaffle traditional jars on the classpath will work fine with the current version of FCS. Modules built for JDK9, and loaded as modules from the module path rather than jars from the classpath, will not be scanned until proper Java 9 support is complete.

**Comment by HoldYourWaffle** [comment]
I'll wait with transitioning to jigsaw then. Thanks!

**Comment by bennylut** [comment]
Hi, Thank you for the great work - is there progress in this issue?

**Comment by lukehutch** [comment]
@bennylut not yet, I haven't found a way to get the Java 9 module API to enumerate the modules in the runtime, and the jars backing each module (or alternatively the files in each module jar). I would appreciate some help, if you have good knowledge of Java 9. See this response I got on jigsaw-dev:

http://mail.openjdk.java.net/pipermail/jigsaw-dev/2017-December/013449.html

**Comment by GedMarc** [comment]
https://docs.oracle.com/javase/9/docs/api/index.html?java/lang/module/package-summary.html
ModuleFinder / ModuleDescripter?

As Java 8 (DDD JVM) / 9 (DDD JDK) is now a Domain Driven Language, Strict encapsulation and Strict Classpath Specifications are priority one. This means that nothing can ever be reflected that isn't exposed (as here https://zeroturnaround.com/wp-content/uploads/2017/11/java-9-modules-visibility-640x228.png), it also means that there is absolutely no reason to ever use an EE container from JDK8 and 9 ever again (hence its cancellation)

**And this is correct!**

I'm not sure if I understood correctly Luke, but similar to using a ServiceLocator to enumerate all the services from a JVM level, perhaps?
I would use the Module Reflector to enumerate all exposed packages and classes, then filter with white listing. Then scan those classes in their given classpath to generate the ClassInfo MethodInfo FieldInfo.?

Fact is, DDD requires that everything be self aware, self exposing, and self-maintained, And the performance benefits that I've felt so far actually reach the level of unbelievable. Much like moving to the Type-Safety engine of JDK7 (and not using the backwards compatibility classes to read JDK6 objects, no performance benefit there, it actually degrades performance drastically when pairing JDK6 classes and JDK7 classes together)

I don't know if this is possible, but it would be great to find out. I think the centralized point is now the OSGI MANIFEST.MF file, which every project after JDK7 should be utilizing, in pairing with module-info from JDK9 you gain a complete view and visibility of each modules classpath and exposed classes.

I must say, It is really good to finally have Java as a DDD supported language!!! instead of people hacking together methods to emulate it.

**Comment by lukehutch** [comment]
@GedMarc have you tried what you suggest? I couldn't get it to work. For all the obstacles I ran into, see my original message to jigsaw-dev, linked above.

**Comment by lukehutch** [comment]
TODO: Test 80 fails when running on JDK9+, because there is no rt.jar, which FCS uses to scan for system classes.

**Comment by GedMarc** [comment]
OK, So I may have actually found some time to be able to do something here...

A few things - Dunno if you want to start a new thread JDK9 is End of Life, so this thread is invalid now. JDK 10.0.1 seems to be it. and JDK 8. I think a new issue? Whatever it is, it needs to support dual compilation xD

Can I switch to JUnit 5? I think this will help a bit with the test cases thing, also please can you make a branch jdk10 or something that I can pull request into separate to the master branch. I can't do what you do so I get the feeling some stuff will need to be double checked.
```
@Test
@EnabledOnJre(JAVA_8)
void onlyOnJava8() {
    // ...
}

@Test
@EnabledOnJre({ JAVA_9, JAVA_10 })
void onJava9Or10() {
    // ...
}

@Test
@DisabledOnJre(JAVA_9)
void notOnJava9() {
    // ...
}
```
I'm going to go src/alternatives route for the dual compilation, and switch between the ScanResultProcessor with a service loader for each Jdk?

**Comment by lukehutch** [comment]
@GedMarc thanks for the offer to help!

I am currently compiling for JDK 7, and there were already several users that complained that JDK 6 wasn't supported. JDK 8 is way too progressive for some people, so I think it will be a few years before everybody gets on the JDK 10+ six-month release cycle bandwagon.

I have been planning to support Java modules via reflection until the vast majority of users are on JDK 10+. I was planning on reflection rather than some other mechanism such as src/alternatives, because there should only be a small number of API calls needed to support module scanning on JDK 10+. Using a ServiceLoader is a good idea, and it would be possible to have two different versions of FCS, but I don't think the extra complexity of src/alternatives would be needed. I'm willing to change this though if it really does seem to be the right thing to do.

It's not ScanResultProcessor that needs to be changed as much as there needs to be a new abstraction of ClasspathElement. Currently there's ClasspathElementDir and ClasspathElementZip. There would need to be a ClasspathElementModule, and the API of ClasspathElement would need to be changed to take into account the different mechanism needed to scan the module path. This change may have to happen in conjunction with changing the ClassLoaderHandler API, since with the module path, you don't fetch a list of jars and dirs to scan from the ClassLoader, you just query the module loader directly as you are scanning.

First things first though, module scanning should be relatively simple on Java 10+ (as much as a few module loader API calls, if it's possible to get it working at all -- from my own investigation so far, I think the API is missing a couple of things, as I described in my post to the JPMS mailing list). In other words, for determining how to do module path scanning, you don't need to experiment on that aspect within the FCS tree, you can do it in any project. Once you or I or somebody else figures out how to do it, we can figure out how to properly restructure FCS to support that scanning mode.

As far as switching to JUnit 5, by all means, please go ahead! Would love some help :-)

**Comment by lukehutch** [comment]
@GedMarc I made a new branch, https://github.com/lukehutch/fast-classpath-scanner/tree/module-path-scanning , as you requested.

**Comment by lukehutch** [comment]
Good news! I finally figured out the API. I was wrong, there is in fact a way to list non-system modules. The problem was that I didn't realize that the API call `classRef.getModule().getLayer().getConfiguration().modules()` returned a `List<ResolvedModule>`, rather than just a `List<Module>`. It turns out that `ResolvedModule` does in fact have a `.reference()` method to get the elusive `ModuleReference` for a module!

This is not documented anywhere that I have seen. It was a royal pain to figure this out. But here is the code that gets a list of system modules from the boot layer, then gets all classes on the call stack, and from each class, finds all modules in the same `Layer`, then gets a `ModuleReference` for each module. With `ModuleReference`, you can easily list files in a module, and open any file.

```java
public class Java9Scanner {

    private static CallerResolver CALLER_RESOLVER;

    static {
        try {
            // This can fail if the current SecurityManager does not allow
            // RuntimePermission ("createSecurityManager"):
            CALLER_RESOLVER = new CallerResolver();
        } catch (final SecurityException e) {
        }
    }

    private static final class CallerResolver extends SecurityManager {
        @Override
        protected Class<?>[] getClassContext() {
            return super.getClassContext();
        }
    }

    // -------------------------------------------------------------------------------------------------------------

    /** Get ModuleReferences from a Layer Configuration. */
    static void findModuleReferences(Configuration layerConfiguration, Set<ModuleReference> moduleReferencesOut) {
        // Get ModuleReferences from layer configuration
        for (ResolvedModule module : layerConfiguration.modules()) {
            ModuleReference ref = module.reference();
            moduleReferencesOut.add(ref);
        }
        // Recurse to parent layer configurations
        List<Configuration> parents = layerConfiguration.parents();
        if (parents != null) {
            for (Configuration parent : parents) {
                findModuleReferences(parent, moduleReferencesOut);
            }
        }
    }

    /** Get ModuleReferences from a Class reference. */
    static void findModuleReferences(Class<?> cls, Set<ModuleReference> moduleReferencesOut) {
        ModuleLayer layer = cls.getModule().getLayer();
        if (layer != null) {
            Configuration cfg = layer.configuration();
            if (cfg != null) {
                findModuleReferences(cfg, moduleReferencesOut);
            }
        }
    }

    // -------------------------------------------------------------------------------------------------------------

    public static void main(String[] args) throws Exception {
        Set<ModuleReference> systemModuleRefs = ModuleFinder.ofSystem().findAll();

        Set<ModuleReference> nonSystemModuleRefs = new HashSet<>();
        if (CALLER_RESOLVER == null) {
            System.out.println(Java9Scanner.class.getSimpleName() + " could not create "
                    + CallerResolver.class.getSimpleName() + ", current SecurityManager does not grant "
                    + "RuntimePermission(\"createSecurityManager\")");
        } else {
            final Class<?>[] callStack = CALLER_RESOLVER.getClassContext();
            if (callStack == null) {
                System.out.println(Java9Scanner.class.getSimpleName() + ": " + CallerResolver.class.getSimpleName()
                        + "#getClassContext() returned null");
            } else {
                for (int i = callStack.length - 1; i >= 0; --i) {
                    System.out.println("Finding module references for " + callStack[i]);
                    findModuleReferences(callStack[i], nonSystemModuleRefs);
                }
            }
        }
        nonSystemModuleRefs.removeAll(systemModuleRefs);

        // Show info for non-system modules
        for (ModuleReference m : nonSystemModuleRefs) {
            System.out.println(m.descriptor().name());
            System.out.println("  Version: " + m.descriptor().toNameAndVersion());
            System.out.println("  Packages: " + m.descriptor().packages());
            Optional<URI> location = m.location();
            if (location.isPresent()) {
                System.out.println("  Location: " + location.get());
            }
            m.open().list().forEach(s -> System.out.println("    File: " + s)); // e.g. "java/util/Scanner.class"
        }
    }
}
```

I'm going to have to test how parallelizable the module system is. I suspect it employs a lot of locking, which will kill performance, at least for scanning system modules. The good news is that for non-system modules, `ModuleReference#location()` returns a "real" URI, of the form `file:///path/to/module.jar`, so I can still scan those in parallel if the module loading system is not as parallelizable as FCS' own scanning logic. However, that will mean I will need to parse all the `module-info.class` files, and replicate all of Java's internal module visibility and export logic, which will be complicated and hard to get right.

(For system modules, the returned URIs are useless, they are just "jrt:/" plus the module name. So I will have to use `ModuleReference#open()#list()` and `ModuleReference#open()#read()` instead of manually scanning.)

**Comment by GedMarc** [comment]
You are the man!!!!

**Comment by GedMarc** [comment]
Can you add an entry into MANIFEST for Automatic-Module-Name :)

Automatic-Module-Name: xxxx.xxxx ?

**Comment by lukehutch** [comment]
Fixed, implementing full JDK9+ module scanning support, and released in version 3.0.0.

**Comment by lukehutch** [comment]
@GedMarc I went one better than adding an Automatic-Module-Name, I discovered this super-cool project that creates a proper module:

https://github.com/moditect/moditect

So FCS can be used by modular projects now too -- it is fully modularized.

What's more, FCS still works on JDK7 and JDK8 projects too, since Moditect uses Objweb-ASM to compile the `module-info.class` (so I was still able to set the source and target level to 1.7 in spite of the project now having a `module-info.class` file).

**Comment by kaqqao** [comment]
Just curious, have you looked into how Spring handles modulepath scanning?

**Comment by seanf** [comment]
@lukehutch Thanks for your work on this! Note that the [wiki](https://github.com/lukehutch/fast-classpath-scanner/wiki) still says "Eventually, the Java 9 module system will be supported (work has not started on this yet -- patches are welcome)."

**Comment by lukehutch** [comment]
@kaqqao I have not seen their code, but I'd like to look at it -- do you know where exactly their module path scanning code is located?

@seanf Thanks, I thought I had caught all of those :-)
